server/common/app_config.py

```diff
@@ -375,6 +389,16 @@ def handle_multi_dataset(self, context):
         if self.multi_dataset__dataroot is None:
             return
 
+        if type(self.multi_dataset__dataroot) == str:
+            self.multi_dataset__dataroot = dict(d=self.multi_dataset__dataroot)
+
+        for key in self.multi_dataset__dataroot.keys():
+            # sanity check for well formed keys
+            if type(key) != str:
```

Comment:
Suggest you add an error check that the key is URL-safe, ie, the key only contains characters that are not reserved for special use in URL.  If you want to be super-duper safe, mandate only characters in the `unreserved` set for all URIs, which are 

   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"

From [RFC 3986](https://www.ietf.org/rfc/rfc3986.txt)

You could probably include a bit more punctuation, but I'm not sure it is all that useful.  For example, JS encodeURI() treats the following as unreserved:  A-Z a-z 0-9 ; , / ? : @ & = + $ - _ . ! ~ * ' ( ) #